src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -28,6 +29,94 @@ static ucs_config_field_t uct_cuda_ipc_md_config_table[] = {
     {NULL}
 };
 
+typedef struct {
+    /* GPU Device number */
+    int     dev_num;
+    /* Cache of accessible devices (ucs_ternary_auto_value_t) */
+    uint8_t accessible[0];
+} uct_cuda_ipc_dev_cache_t;
+
+static UCS_F_ALWAYS_INLINE int uct_cuda_ipc_uuid_equals(CUuuid a, CUuuid b)
+{
+    int64_t *a64 = (int64_t *)a.bytes;
+    int64_t *b64 = (int64_t *)b.bytes;
+    return (a64[0] == b64[0]) && (a64[1] == b64[1]);
+}
+
+static UCS_F_ALWAYS_INLINE khint32_t uct_cuda_ipc_uuid_hash_func(CUuuid key)
+{
+    int64_t *i64 = (int64_t *)key.bytes;
+    return (khint32_t)(i64[0]);
+}
+
+KHASH_INIT(cuda_ipc_uuid_hash, CUuuid, uct_cuda_ipc_dev_cache_t*, 1,
+           uct_cuda_ipc_uuid_hash_func, uct_cuda_ipc_uuid_equals);
+
+typedef struct {
+    khash_t(cuda_ipc_uuid_hash) hash;
+    ucs_recursive_spinlock_t    lock;
+} uct_cuda_ipc_uuid_cache_t;
+
+static uct_cuda_ipc_uuid_cache_t uct_cuda_ipc_uuid_cache;
+
+static uct_cuda_ipc_dev_cache_t *uct_cuda_ipc_create_dev_cache(int dev_num)
+{
+    uct_cuda_ipc_dev_cache_t *cache;
+    ucs_status_t status;
+    int i, num_devices;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGetCount(&num_devices));
+    if (UCS_OK != status) {
+        ucs_error("cuDeviceGetCount() failed: %s", ucs_status_string(status));
+        return NULL;
+    }
+
+    cache = ucs_malloc(sizeof(*cache) + num_devices,
+                       "uct_cuda_ipc_dev_cache_t");
+    if (cache == NULL) {
+        ucs_error("failed to allocate memory for uct_cuda_ipc_dev_cache_t");
+        return NULL;
+    }
+
+    cache->dev_num = dev_num;
+    for (i = 0; i < num_devices; ++i) {
+        cache->accessible[i] = UCS_TRY;
+    }
+
+    return cache;
+}
+
+static uct_cuda_ipc_dev_cache_t *uct_cuda_ipc_get_dev_cache(const CUuuid *uuid)
+{
+    uct_cuda_ipc_dev_cache_t *cache   = NULL;
+    khash_t(cuda_ipc_uuid_hash) *hash = &uct_cuda_ipc_uuid_cache.hash;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(cuda_ipc_uuid_hash, hash, *uuid, &ret);
+    switch (ret) {
+    case UCS_KH_PUT_BUCKET_EMPTY:
+    case UCS_KH_PUT_BUCKET_CLEAR:
+        cache = uct_cuda_ipc_create_dev_cache(kh_size(hash) - 1);
+        if (NULL != cache) {
+            kh_val(hash, iter) = cache;
+        } else {
+            kh_del(cuda_ipc_uuid_hash, hash, iter);
```

Comment:
By this point in time we have added a new key in the hashmap with empty value.
Now we have 2 options with subtle difference in behavior:
1) we can either remove from the map a pair {key -> NULL}, which I think is more logical. The important point is that by removing we also decrease the hashmap size, which is used to assing unique ID for dev_num (see line 100). This way we maintain in the hashmap only entries with valid unique IDs.
2) we can keep pair {key -> NULL} in the map.
It will have some side effect on unique ID assignment: we might have gaps in the IDs -> suboptimal GPU distribution across the streams. On the other side, next time someone tries to create a new entry with this UUID, we take a null entry from cache and don't even try to create.

I thought that p1 is desirable behaviour, therefore this logic. If you think p2 is the way to go - please let me know. Not a big deal anyway

Reply:
i think p2 is less code and we won't try to add the same uuid over and over again

Reply:
done